packages/dht/src/connection/ManagedConnection.ts

```diff
@@ -0,0 +1,398 @@
+import { ConnectionEvents, ConnectionID, ConnectionType, IConnection } from "./IConnection"
+import * as Err from '../helpers/errors'
+import { Handshaker } from "./Handshaker"
+import { PeerDescriptor } from "../proto/packages/dht/protos/DhtRpc"
+import { Logger, raceEvents3, runAndRaceEvents3, RunAndRaceEventsReturnType } from "@streamr/utils"
+import EventEmitter from "eventemitter3"
+import { PeerIDKey } from "../helpers/PeerID"
+import { keyFromPeerDescriptor } from '../helpers/peerIdFromPeerDescriptor'
+import { DisconnectionType } from "../transport/ITransport"
+
+export interface ManagedConnectionEvents {
+    managedData: (bytes: Uint8Array, remotePeerDescriptor: PeerDescriptor) => void
+    handshakeRequest: (peerDescriptor: PeerDescriptor) => void
+    handshakeCompleted: (peerDescriptor: PeerDescriptor) => void
+    handshakeFailed: () => void
+    bufferSentByOtherConnection: () => void
+    closing: () => void
+    internal_disconnected: () => void
+}
+
+const logger = new Logger(module)
+
+export type Events = ManagedConnectionEvents & ConnectionEvents
+export class ManagedConnection extends EventEmitter<Events> {
+
+    private static objectCounter = 0
+    public objectId = 0
+    private implementation?: IConnection
+
+    private outputBuffer: Uint8Array[] = []
+    private inputBuffer: [data: Uint8Array][] = []
```

Comment:
Is this an array of arrays? I.e. there are multiple input buffers? (if so, rename to `inputBuffers`?)